WIP: Downgrade schema v5 to v4

winksaville opened the pull request.

winksaville committed:
WIP: Downgrade schema v5 to v4

winksaville commented:
@michaelsproul, initial stab at downgrade, still need to add the `db downgrade` cli command, but would appreciate it if you could spare a few minutes to catch obvious errors.

michaelsproul commented on beacon_node/beacon_chain/src/schema_change.rs:
```diff
@@ -93,6 +94,47 @@ pub fn migrate_schema<T: BeaconChainTypes>(
+                            None => (),
+                        }
+                    }
+                    Err(_) => (),
```
We should error if the store errors, rather than failing silently, i.e. match on `db.load_anchor_info()?`

winksaville replied:
Done

michaelsproul reviewed (commented):
Looking good!

Just one suggestion

Thanks for taking the time to implement this :blush:

michaelsproul commented on beacon_node/store/src/hot_cold_store.rs:
```diff
@@ -72,7 +72,7 @@ pub struct HotColdDB<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> {
     pub(crate) spec: ChainSpec,
     /// Logger.
-    pub(crate) log: Logger,
+    pub log: Logger,
```
This is pretty benign, so I'm happy to make this field public

michaelsproul replied:
The alternative would be injecting the logger here:

https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/store/src/hot_cold_store.rs#L157

But that doesn't seem any more elegant

winksaville replied:
Leaving as is for now, we can change it later if we want.

winksaville committed:
Initial stab at a database_manager UI

winksaville commented:
@michaelsproul , I added commit "Initial stab at a database_manager Ui", when you have time please take look.

Questions:
 - Is the sub-command `database_manager` OK?
 - Is `open_as_is` the right direction for "opening" the database?
 - What do you advise is the proper way to  initialize the parameters to `open_as_is`?

Txs!

michaelsproul commented:
> Is the sub-command database_manager OK?

Yep! We can alias it easily enough at the CLI to `lighthouse db` if we want

> Is open_as_is the right direction for "opening" the database?

Yeah, although you could keep `open` the same and just pass a no-op schema upgrade like `|_, _, _| Ok(())`.

> What do you advise is the proper way to initialize the parameters to open_as_is?

This one's a bit trickier. Currently the DB is only opened from the BN subcommand, so its config file parsing is embedded into the BN config, e.g.  here: https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/client/src/config.rs#L51-L54

To avoid getting entangled in that I think we could have 3 separate flags for the `database_manager`:

* `--datadir` for setting the datadir. The best utility for determining this is `get_data_dir` in the BN directory, so you may need to make `database_manager` depend on `beacon_node`. Set the hot DB path to `$datadir/beacon/chain_db`.
* `--freezer-dir` for setting the freezer DB. If not provided it should default to `$datadir/beacon/freezer_db`.
* `--slots-per-restore-point`, default 2048.

The `StoreConfig` can be constructed with default values for its other fields (they're not really relevant):

```rust
let store_config = StoreConfig {
    slots_per_restore_point, /* from CLI or 2048 */
    ..StoreConfig::default()
};
```

Finally, for the `ChainSpec`, that's best sourced from a network configuration. So we should have a `--network` flag, which can be parsed by `get_eth2_network_config` (also in `beacon_node`). Given a `NetworkConfig` we can get the `ChainSpec` using `chain_spec`:

https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/common/eth2_network_config/src/lib.rs#L88-L96

Sorry, I know that's a lot! Hope that helps!

winksaville commented on database_manager/src/lib.rs:
```diff
@@ -0,0 +1,67 @@
+        HotColdDB::open_as_is(hot_path.as_path(), cold_path.as_path(), config, spec, log)?;
+    let schema_version = db.load_schema_version()?;
+
+    println!("database version: {:?}", schema_version);
```
@michaelsproul have I done something wrong, why is schema_version `None`?
```
wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Downgrade-schema-v5-to-v4)
$ cargo run --bin lighthouse db -v
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `target/debug/lighthouse db -v`
Running database manager for mainnet network
database version: None
```

michaelsproul replied:
The code looks OK to me. Are you running this against a database that already exists?

winksaville replied:
No, it's a new database in a the tempdir.

michaelsproul replied:
That makes sense then, open_as_is never writes the schema version because that happens when the migration runs.

winksaville replied:
@michaelsproul I've got a solution, open_as_is now returns db and schema version. Does that work for you?

winksaville committed:
Change open_as_is to set schema if new database

It now returns both the db and and its schema_version.

Now we get the expected schema_version:
```
wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Downgrade-schema-v5-to-v4)
$ cargo run --bin lighthouse db -v
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/lighthouse db -v`
Running database manager for mainnet network
database version: SchemaVersion(5)
```

winksaville committed:
First attempt at downgrade_db.

Cannot figure out how to invoke migrate_schema as it's a generic function
that must implement the `pub trait BeaconChainTypes`. The two only places
where migrate_schema is called are:

 - Recursively in the implementation itself in
   beacon_node/beacon_chain/src/schema_change.rs

 - beacon_node/store/srch/hot_cold_store.rs::open where it's passed
   as a fn parameter in a call from
   beacon_node/client/src/builder.rs::disk_store

Example:
```
  wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Downgrade-schema-v5-to-v4)
  $ cargo run --bin lighthouse db dg
      Finished dev [unoptimized + debuginfo] target(s) in 0.20s
       Running `target/debug/lighthouse db dg`
  Running database manager for mainnet network
  downgrade database version 5 to 4
```

winksaville commented on database_manager/src/lib.rs:
```diff
@@ -52,18 +136,36 @@ pub fn display_db_version<E: EthSpec>(
+    //    beacon_node/client/src/builder.rs::disk_store
+
+    //migrate_schema::Witness<TSlotClock, TEth1Backend, TEthSpec, THotStore, TColdStore>(_db, datadir.path(), from, to)?;
+    //migrate_schema::<_>(_db, datadir.path(), from, to)?;
```
@michaelsproul, started working on `downgrade` and I can't figure out how to call `migrate_schema`. needs "Witness" magic which is way beyond my rust knowledge :)

michaelsproul replied:
Ah yeah this is a little bit clunky for our purposes but we can work around it by providing (arbitrary) types for the `SlotClock` and `Eth1Chain` like this:

```rust
migrate_schema::<Witness<SystemTimeSlotClock, CachingEth1Backend<E>, E, LevelDB<E>, LevelDB<E>>(..)
```

Type inference should be able to work out the rest if we just hint the first two:

```rust
migrate_schema::<Witness<SystemTimeSlotClock, CachingEth1Backend<E>, _, _, _>(..)
```

`Witness`, `SystemTimeSlotClock` and `CachingEth1Backend` will also need to be imported from their respective crates. Let us know if you have issues

winksaville replied:
@michaelsproul , it is now "working", TYVM!

Could you provide a link or background information on what is going on and why this works?

I saw that @paulhauner added this code in https://github.com/sigp/lighthouse/pull/542 and the indirect call in open was so there wouldn't be a circular code. But I didn't see anything explaining the need for a `Witness` other than in the code we see:
```
/// An empty struct used to "witness" all the `BeaconChainTypes` traits. It has no user-facing
/// functionality and only exists to satisfy the type system.
pub struct Witness<TSlotClock, TEth1Backend, TEthSpec, THotStore, TColdStore>(
    PhantomData<(TSlotClock, TEth1Backend, TEthSpec, THotStore, TColdStore)>,
);
```
Which doesn't help me with why we have to go to a lot of trouble to convince the type system this call is Ok.

michaelsproul replied:
It's all in aid of the [`BeaconChainBuilder`](https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/beacon_chain/src/builder.rs#L57-L65), which allows us to build `BeaconChain`s instantiated with various concrete types for the different components (the slot clock, the eth1 backend, the database, and most importantly the `EthSpec` which chooses between minimal and mainnet presets).

In the builder's case we don't need to specify the witness type manually because it can be inferred by the compiler when the builder methods are called. E.g. the [`testing_slot_clock`](https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/beacon_chain/src/builder.rs#L849-L873) method is specialised to `TSlotClock = TestingSlotClock` so if the compiler sees a `builder.testing_slot_clock(..)` call it can infer that the slot clock type is `TestingSlotClock`.

The reason `migrate_schema` is a bit unwieldy is because it was moved from within the `store` crate (where it didn't care about witnesses) to the `beacon_chain` crate, in order to get access to the `ValidatorPubkeyCache`. For better or for worse, the `ValidatorPubkeyCache` is parametrised by `<T: BeaconChainTypes>`, so that ripples up to the top-level. Until now `migrate_schema` was only called from the builder/beacon chain where a witness was already known.

I think the manual type hint is an OK solution for now, as it avoids us having to refactor (unrelated) code in the `ValidatorPubkeyCache` and gives `migrate_schema` the flexibility to interact with other types parametrised by a witness in future.

winksaville replied:
I agree it's.OK for now, but it's the first time that neither me, rust-analyzer or rustc could figure out the right type hints. Which really makes me feel something isn't right. Probably with more rust experience I could have figured it out, but if a refactoring would make it more obvious I'd say it would be worth it. But, again, I agree not in this PR.

winksaville committed:
Add a_logger and a_null_logger to common/logging

These are unconditional and maybe used without needing having to
declare using a feature.

winksaville committed:
Successfully invoked migrate_schema

Thanks to Michael Sproul for providing the necessary type information
it now "works":
```
wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Downgrade-schema-v5-to-v4)
$ cargo run --bin lighthouse db dg
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `target/debug/lighthouse db dg`
Running database manager for mainnet network
downgrade database version 5 to 4
Oct 13 22:21:36.313 INFO migrate_schema: from=5 to=4, module: beacon_chain::schema_change:25
Oct 13 22:21:36.313 INFO Dowgrade v5 to v4, module: beacon_chain::schema_change:100
Oct 13 22:21:36.313 INFO Dowgrade v5 to v4: db.store_schema_version=4, module: beacon_chain::schema_change:127
Oct 13 22:21:36.313 INFO Dowgrade v5 to v4 completed, module: beacon_chain::schema_change:130
```

michaelsproul commented:
Thanks for the great work on this @winksaville. I've implemented an updated version of the `database_manager` with some more features in a new PR (https://github.com/sigp/lighthouse/pull/3129) and we'll have some DB downgrades built atop that coming soon!

I'll close this PR in favour of the new one
